Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,199 @@
+  exact id (exists_eq_pow_rootMultiplicity_mul_and_not_dvd p hp 1).choose_spec.1.symm
+
+omit [NeZero (1 : R)] [NoZeroDivisors R] in
+theorem greatestFactorOneSubNotDvd_ne_zero :
+    greatestFactorOneSubNotDvd p hp ≠ 0 := fun h0 => by
+  let hpow := pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq p hp
+  rw [h0, mul_zero] at hpow; exact hp <| id hpow.symm
+
+theorem natDegree_greatestFactorOneSubNotDvd_le :
+    (greatestFactorOneSubNotDvd p hp).natDegree ≤ p.natDegree := by
+  have : p.natDegree = ((1 - X : R[X]) ^ p.rootMultiplicity 1 * gFOSND p hp).natDegree := by
+    rw [pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq]
+  rw [this, natDegree_mul]
+  exact (gFOSND p hp).natDegree.le_add_left (natDegree ((1 - X) ^ p.rootMultiplicity 1))
+  exact pow_ne_zero _ <| fun h0 => by
+    let this : (1 - X : R[X]).coeff 0 = 0 := by rw [h0]; simp only [coeff_zero];
+    simp only [coeff_sub, coeff_one_zero, coeff_X_zero, sub_zero, one_ne_zero] at this
+  exact greatestFactorOneSubNotDvd_ne_zero p hp
+
+theorem natDegree_pow_rootMultiplicity_sub_mul_greatestFactorOneSubNotDvd_le
+    (hp1 : d + 1 ≤ p.rootMultiplicity 1) :
+    ((1 - X) ^ ((p.rootMultiplicity 1) - (d + 1)) * greatestFactorOneSubNotDvd p hp).natDegree
+    ≤ p.natDegree := by
+  let this := pow_ne_zero (p.rootMultiplicity 1 - (d + 1)) <| fun (h0 : (1 - X : R[X]) = 0) => by
+    let this : (1 - X : R[X]).coeff 0 = 0 := by rw [h0, coeff_zero];
+    simp only [coeff_sub, coeff_one_zero, coeff_X_zero, sub_zero, one_ne_zero] at this
+  rw [show p.natDegree = (((1 - X : R[X]) ^ (p.rootMultiplicity 1 - (d + 1) + (d + 1))) *
+    gFOSND p hp).natDegree by rw [← Nat.eq_add_of_sub_eq hp1 rfl,
+    pow_rootMultiplicity_mul_greatestFactorOneSubNotDvd_eq], pow_add, mul_assoc,
+    mul_comm ((1 - X) ^ (d + 1)), ← mul_assoc, natDegree_mul, natDegree_mul, natDegree_mul]
+  simp only [natDegree_pow, le_add_iff_nonneg_right, zero_le]
+  · exact this
+  · exact greatestFactorOneSubNotDvd_ne_zero p hp
+  · rw [mul_ne_zero_iff]; exact ⟨this, greatestFactorOneSubNotDvd_ne_zero p hp⟩
+  · exact pow_ne_zero _ <| fun h0 => by
+      let this : (1 - X : R[X]).coeff 0 = 0 := by rw [h0, coeff_zero];
+      simp only [coeff_sub, coeff_one_zero, coeff_X_zero, sub_zero, one_ne_zero] at this
+  · exact this
+  · exact greatestFactorOneSubNotDvd_ne_zero p hp
+
+end greatestFactorOneSubNotDvd
+
+/--
+A polynomial which makes it easier to define the Hilbert polynomial. Look at the theorem
+`Polynomial.preHilbert_eq_choose_sub_add`, which states that for any `d k n : ℕ` with `k ≤ n`,
+`(Polynomial.preHilbert d k).eval (n : ℚ) = Nat.choose (n - k + d) d`.
+-/
+noncomputable def preHilbert (d k : ℕ) : ℚ[X] :=
+  (d.factorial : ℚ)⁻¹ • (∏ i : Finset.range d, (X - (C (k : ℚ)) + (C (i : ℚ)) + 1))
+
+local notation "gFOSND" => greatestFactorOneSubNotDvd
+
+theorem preHilbert_eq_choose_sub_add (d k n : ℕ) (hkn : k ≤ n):
+    (preHilbert d k).eval (n : ℚ) = Nat.choose (n - k + d) d := by
```

Comment:
This proof feels far too long (although maybe I'm just being optimistic) but my first reaction is to wonder whether you really want a variable n here with k<=n, Why not have a natural variable t and let n=k+t? Then you don't have the inequality, you don't have natural subtraction, the result is mathematically equivalent, and probably the proof can get shorter.